core/net/mac/tsch/sixtop/sixtop.c

```diff
@@ -0,0 +1,204 @@
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * \addtogroup sixtop
+ * @{
+ */
+/**
+ * \file
+ *         6TiSCH Operation Sublayer (6top)
+ *
+ * \author
+ *         Yasuyuki Tanaka <yasuyuki.tanaka@inf.ethz.ch>
+ */
+
+#include "lib/assert.h"
+
+#include "net/packetbuf.h"
+#include "net/mac/frame802154.h"
+#include "net/mac/frame802154e-ie.h"
+
+#include "sixtop.h"
+#include "sixp.h"
+
+#define DEBUG DEBUG_PRINT
+#include "net/net-debug.h"
+
+#if NETSTACK_MAC != tschmac_driver
+#error sixtop needs tschmac_driver to be set to NETSTACK_MAC
+#endif
+
+const sixtop_sf_t *schedule_functions[SIXTOP_MAX_SCHEDULE_FUNCTIONS];
+
+/*---------------------------------------------------------------------------*/
+int
+sixtop_add_sf(const sixtop_sf_t *sf)
+{
+  int i;
+
+  assert(sf != NULL);
+
+  if(sixtop_find_sf(sf->sfid) != NULL) {
+    PRINTF("6top: there is another SF installed with the same sfid [%u]\n",
+           sf->sfid);
+    return -1;
+  }
+
+  for(i = 0; i < SIXTOP_MAX_SCHEDULE_FUNCTIONS; i++) {
+    if(schedule_functions[i] == NULL) {
+      schedule_functions[i] = sf;
+      if(sf->init != NULL) {
+        sf->init();
+      }
+      break;
+    }
+  }
+
+  if(i == SIXTOP_MAX_SCHEDULE_FUNCTIONS) {
+    PRINTF("6top: no room to install the specified SF [sfid=%u]\n",
+           sf->sfid);
+    return -1;
+  }
+
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+const sixtop_sf_t *
+sixtop_find_sf(uint8_t sfid)
+{
+  int i;
+
+  for(i = 0; i < SIXTOP_MAX_SCHEDULE_FUNCTIONS; i++) {
+    if(schedule_functions[i] != NULL &&
+       schedule_functions[i]->sfid == sfid) {
+      return (const sixtop_sf_t *)schedule_functions[i];
+    }
+  }
+
+  return NULL;
+}
+/*---------------------------------------------------------------------------*/
+void
+sixtop_output(const linkaddr_t *dest_addr, mac_callback_t callback, void *arg)
+{
+  assert(dest_addr != NULL);
+
+  packetbuf_set_addr(PACKETBUF_ADDR_RECEIVER, dest_addr);
+  packetbuf_set_addr(PACKETBUF_ADDR_SENDER, &linkaddr_node_addr);
+  NETSTACK_MAC.send(callback, arg);
+}
+/*---------------------------------------------------------------------------*/
+void
+sixtop_input(void (*llsec_input)(void))
+{
+  uint8_t *hdr_ptr, *payload_ptr;
+  uint16_t hdr_len, payload_len;
+
+  frame802154_t frame;
+  struct ieee802154_ies ies;
+  linkaddr_t src_addr;
+
+  assert(llsec_input != NULL);
+
+  /*
+   * A received *DATA* frame is supposed to be stored in packetbuf by
```

Comment:
Thanks, this is what I forgot to ask of the 6TiSCH WG.

The current implementation could handle a 6P message, for example, in an EACK frame. But, I think it's not intended by the 6P draft. Any type of frame can have Payload IE. So, I think we need to define which type of frame can have an IETF IE.